great_expectations/data_context/types/base.py

```diff
@@ -15,88 +20,536 @@
+        ]
+    return substitute_config_variable(data, replace_variables_dict)
+
+
+def object_to_yaml_str(obj):
+    output_str: str
+    with StringIO() as string_stream:
+        yaml.dump(obj, string_stream)
+        output_str = string_stream.getvalue()
+    return output_str
 
 
-class DataContextConfig(DictDot):
+class BaseConfig(DictDot):
+    def __init__(
+        self, commented_map: CommentedMap = None,
+    ):
+        if commented_map is None:
+            commented_map = CommentedMap()
+        self._commented_map = commented_map
+
+    @classmethod
+    def from_commented_map(cls, commented_map: CommentedMap):
+        raise NotImplementedError
+
+    @property
+    def commented_map(self) -> CommentedMap:
+        return self._commented_map
+
+    def get_schema_validated_updated_commented_map(self) -> CommentedMap:
+        raise NotImplementedError
+
+    def to_yaml(self, outfile):
+        yaml.dump(self.get_schema_validated_updated_commented_map(), outfile)
+
+    def to_yaml_str(self) -> str:
+        return object_to_yaml_str(self.get_schema_validated_updated_commented_map())
+
+    def to_dict(self) -> dict:
+        return dict(self.get_schema_validated_updated_commented_map())
+
+
+class DataContextConfig(BaseConfig):
+    # noinspection SpellCheckingInspection
     def __init__(
         self,
         config_version,
-        datasources,
-        expectations_store_name,
-        validations_store_name,
-        evaluation_parameter_store_name,
-        plugins_directory,
-        validation_operators,
-        stores,
-        data_docs_sites,
+        datasources=None,
+        expectations_store_name=None,
+        validations_store_name=None,
+        evaluation_parameter_store_name=None,
+        plugins_directory=None,
+        validation_operators=None,
+        stores=None,
+        data_docs_sites=None,
         notebooks=None,
         config_variables_file_path=None,
         anonymous_usage_statistics=None,
         commented_map=None,
     ):
-        if commented_map is None:
-            commented_map = CommentedMap()
-        self._commented_map = commented_map
         self._config_version = config_version
         if datasources is None:
             datasources = {}
-        self.datasources = datasources
-        self.expectations_store_name = expectations_store_name
-        self.validations_store_name = validations_store_name
-        self.evaluation_parameter_store_name = evaluation_parameter_store_name
+        self._datasources = datasources
+        self._expectations_store_name = expectations_store_name
+        self._validations_store_name = validations_store_name
+        self._evaluation_parameter_store_name = evaluation_parameter_store_name
         self.plugins_directory = plugins_directory
-        if not isinstance(validation_operators, dict):
+        if validation_operators is None:
+            validation_operators = {}
+        elif not isinstance(validation_operators, dict):
```

Comment:
I thought the same, but did not want to do it in this PR, because I did not have the time to find out why it was done selectively.  We can pursue this in a subsequent PR.